src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,126 @@
+    private static final double GROWTH_FACTOR = 1.5;
+
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        ensureCapacity();
+
+        if (index < size) {
+            System.arraycopy(elements, index, elements, index + 1, size - index);
+        }
+        elements[index] = value;
+        size++;
 
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elements[index];
+    }
+
+    private int getIndex(T value) {
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(value, elements[i])) {
+                return i;
+            }
+        }
+        return -1;
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T currentElement = get(index);
+        System.arraycopy(elements, index + 1, elements, index, elements.length - (index + 1));
+        size--;
+        return currentElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getIndex(element);
+        if (index < 0) {
+            throw new NoSuchElementException("Element not found: " + element);
+        }
+
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private void ensureCapacity() {
+        if (size == elements.length) {
+            grow();
+        }
+    }
+
+    private void grow() {
+        int newCapacity = (int) (size * GROWTH_FACTOR);
+        T[] newArray = (T[]) new Object[newCapacity];
+        System.arraycopy(elements, 0, newArray, 0, size);
+        elements = newArray;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds for length " + size);
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "ArrayList{"
+                + ", elements=" + Arrays.toString(elements)
```

Comment:
The 'toString' method includes ', elements=' which is misleading since there's no preceding element in the string. You should remove the leading comma.